File: dmoj/riceball.cpp
#include <bits/stdc++.h>
using namespace std;
int N, balls[410], solved[410][410];

int canMakeBall(int s, int e){
    if(s >= e){
        return 1;
    }
    if(solved[s][e] != -1){
        return solved[s][e];
    }
    for(int i=s; i<=e; i++){
        for(int j=i+1; j<=e; j++){
            int tot1=0, tot2=0;
            for(int k=s; k<=i; k++){
                tot1 += balls[k];
            }
            for(int k=j; k < min(N, e+1); k++){
                tot2 += balls[k];
            }
            if(tot1 == tot2){
                if(canMakeBall(s,i) * canMakeBall(i+1, j-1) * canMakeBall(j,e) == 1){
                    solved[i][j] = 1;
                    return 1;
                }
            }
        }
    }
    solved[s][e] = 0;
    return 0;
}
int main(){
    cin.sync_with_stdio(0); cin.tie(0);
    cin >> N;
    memset(solved, -1, sizeof solved);
    int best = 0;
    for(int i=0; i<N; i++){
        cin >> balls[i];
        solved[i][i] = 1;
        best = max(best, balls[i]);
    }
    for(int i=0; i<N; i++){
        for(int j=i+1; j<N; j++){
            if(canMakeBall(i,j)){
                int sum = 0;
                for(int k=i; k<=j; k++){
                    sum += balls[k];
                }
                if(sum > best){
                    best = sum;
                }
            }
        }
    }
    cout << best << endl;
}

File: dmoj/riceball.py
import sys
input = sys.stdin.readline
num = int(input())
balls = [int(i) for i in input().split()]

solved = [[-1 for i in range(num)] for i in range(num)]
for i in range(num):
    solved[i][i] = 1

def canMakeBall(start, end):
    if start>=end:
        return 1
    if solved[start][end] != -1:
        return solved[start][end]
    for i in range(start,end+1):
        for j in range(i+1,end+1):
            
            if sum(balls[start:i+1]) == sum(balls[j:min(len(balls), end+1)]):
                
                if canMakeBall(start,i) * canMakeBall(i+1,j-1) * canMakeBall(j,end) == 1:
                    solved[i][j] = 1
                    return 1
    solved[start][end] = 0
    return 0

best = max(balls)
for i in range(num):
    for j in range(i+1,num):
        if canMakeBall(i,j):
            if sum(balls[i:j+1]) > best:
                best = sum(balls[i:j+1])
print(best)
